design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,217 @@
+cluster. Assuming you have installed Rook and the corresponding CockroachDB
+operator into a cluster, you should be able to dynamically provision a
+CockroachDB cluster into the Kubernetes cluster from your Crossplane control
+cluster. The experience would look as follows:
+
+1. Create a non-portable Rook CockroachDB `ClusterClass`
+
+```yaml
+apiVersion: cockroachdb.rook.crossplane.io/v1alpha1
+kind: ClusterClass
+metadata:
+  name: rook-cockroachdb
+  namespace: rook-infra-dev
+specTemplate:
+  clusterSelector:
+    matchLabels:
+        app: my-cool-cluster
+  # full documentation on all available settings can be found at:
+  # https://rook.io/docs/rook/master/cockroachdb-cluster-crd.html
+  scope:
+    nodeCount: 3
+    # You can only have one PersistentVolumeClaim in this list!
+    volumeClaimTemplates:
+    - metadata:
+        name: rook-cockroachdb-data
+      spec:
+        accessModes: [ "ReadWriteOnce" ]
+        # Uncomment and specify your StorageClass, otherwise
+        # the cluster admin defined default StorageClass will be used.
+        #storageClassName: "your-cluster-storageclass"
+        resources:
+          requests:
+            storage: "1Gi"
+  network:
+    ports:
+    - name: http
+      port: 8080
+    - name: grpc
+      port: 26257
+  secure: false
+  cachePercent: 25
+  maxSQLMemoryPercent: 25
+  # A key/value list of annotations
+  annotations:
+  #  key: value
+```
+
+2. Create a portable `PostgreSQLInstanceClass`
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstanceClass
+metadata:
+  name: postgresql-standard
+  namespace: app-project1-dev
+classRef:
+  kind: ClusterClass
+  apiVersion: cockroachdb.rook.crossplane.io/v1alpha1
+  name: rook-cockroachdb
+  namespace: rook-infra-dev
+```
+
+3. Create a `PostgreSQLInstance` claim
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstance
+metadata:
+  name: app-postgresql
+  namespace: app-project1-dev
+spec:
+  classRef:
+    name: postgresql-standard
+  writeConnectionSecretToRef:
+    name: postgresqlconn
+  engineVersion: "9.6"
+```
+
+On creation of the `PostgreSQLInstance` claim, the claim controller would
+configure and create a `Cluster` (`cockroachdb.rook.crossplane.io/v1alpha1`)
+instance. Then, the managed controller would use the Crossplane `Cluster`
+instance to create a `Cluster` (`cockroachdb.rook.io/v1alpha1`) in the target
+Kubernetes cluster where Rook and the CockroachDB operator are installed. This
+workflow is similar to dynamic provisioning in a typical infrastructure
+provider:
+
+![K8s Native Providers](./images/k8s-native-providers.png)
+
+## Technical Design
+
+### Claim Reconciler
+
+The claim controllers could use the same shared [claim reconciler] that other
+existing infrastructure stacks use currently, providing their own
+`ManagedConfigurators` to configure the managed resource instance using the
+referenced class.
+
+### Scheduler Reconciler
```

Comment:
The purpose of the scheduler is to decide which remote cluster to use for creation of the "external resources". Much like `KubernetesApplication`, this will usually involve just getting the single cluster referenced using the `clusterSelector` field, but could involve round-robin scheduling if multiple clusters match selectors. I don't think we necessarily *need* this functionality, so I would be open to removing.

Reply:
But we have only 1 remote cluster that we get the kubeconfig of during the connector phase. I am not sure we want to support more than one Rook cluster that we can provision resources on at this phase. The current support for that kind of setup is simply having another `Provider` resource that resource classes refer to statically.

Reply:
Important note here is that the managed reconciler would not accept a resource that had not already been scheduled, so by the time the resource gets to the managed reconciler, there will only be one kubeconfig that it needs to obtain in the connector phase.

Reply:
I'm a little ambivalent around whether we need a scheduler here. It's useful for all the reasons being able to schedule a workload is useful, but I lean slightly toward the simpler and faster path that omits it to begin with.

Reply:
if there's no scheduling capability, how would the reconciler know what target cluster to propagate the resource to?  Say there are 2 `KubernetesCluster` resources known to Crossplane...it has to pick one somehow, does it?  Perhaps the naive `clusterSelector` that @hasheddan mentioned is a cheap initial solution to this.

Reply:
simpler, faster path 👍

Reply:
@jbw976 I guess the alternative would be to have an object reference to a `KubernetesCluster` that we set directly, rather than labels that the scheduler uses to find and set a reference to a cluster.

Reply:
I don't know that we should take it for granted that we want to target `KubernetesCluster` here. Remember that every other managed resource uses a `Provider` for this case, and that we have no BYO `KubernetesCluster` or "use the `KubernetesCluster` you're running on" story.

Reply:
I think if we don't have a good reason for a scheduler between `Provider` objects in other infra-stacks, we don't have one for Rook, too.

> Remember that every other managed resource uses a Provider for this case

@negz As far as I see, the only difference for Rook should be about the content of that `Provider` object. It will be a kubeconfig instead of username/password or a json. If the user wants to use `KubernetesCluster` as a Rook cluster, they can just put its kubeconfig in a `Rook Provider` object and it'd work. I agree that we don't need to know whether it's `KubernetesCluster` or some cluster we have no idea of or even control cluster itself.

> if there's no scheduling capability, how would the reconciler know what target cluster to propagate the resource to?

@jbw976 Rook Provider object that is referred as `providerRef` in the resource would have the right kubeconfig to provision that resource.

Reply:
Discussion of scheduling has moved to future considerations, so I'm going to resolve this thread.